### ISSUE TITLE
Support preprocessing channels in ::http

### ISSUE THREAD

**Issue by dbohdan** [dbohdan/jimhttp#8]

**Comment by dimkr** [comment]
Sample use case: wrapping client sockets with TLS

``` diff
diff --git a/http.tcl b/http.tcl
index 2177d68..8ea391b 100644
--- a/http.tcl
+++ b/http.tcl
@@ -287,6 +287,10 @@ proc ::http::serve-and-trap-errors {channel clientAddr clientPort routes} {
 # Handle HTTP requests over a channel and send responses. A hacky HTTP
 # implementation.
 proc ::http::serve {channel clientAddr clientPort routes} {
+    # hack
+    tls load_cert ../rootCA.pem
+
+    set schannel [$channel tls -server]
     global ::http::maxRequestLength

     ::http::log info "Client connected: $clientAddr"
@@ -295,7 +299,8 @@ proc ::http::serve {channel clientAddr clientPort routes} {

     set headerLines {}
     set firstLine 1
-    while {[gets $channel buf]} {
+    puts $schannel
+    while {[gets $schannel buf]} {
         if {$firstLine} {
             # Change the newline variable when the incoming request has
             # nonstandard \n newlines. This happens, e.g., when you use netcat.
@@ -313,6 +318,9 @@ proc ::http::serve {channel clientAddr clientPort routes} {
             break
         }
         lappend headerLines $buf
+
+        # hack, TLS socket support doesn't handle \n yet
+        break
     }

     set request [::http::parse-headers $headerLines]
@@ -333,10 +341,10 @@ proc ::http::serve {channel clientAddr clientPort routes} {
             if {$request(contentLength) <= $::http::maxRequestLength} {
                 if {[dict exists $request expect] &&
                             ($request(expect) eq "100-continue")} {
-                    puts $channel "HTTP/1.1 100 Continue\n"
+                    puts $schannel "HTTP/1.1 100 Continue\n"
                 }

-                set postString [read $channel $request(contentLength)]
+                set postString [read $schannel $request(contentLength)]
                 if {$request(contentType) eq
                         "application/x-www-form-urlencoded"} {
                     ::http::log debug "POST request: {$postString}\n"
@@ -386,11 +394,12 @@ proc ::http::serve {channel clientAddr clientPort routes} {

     if {!$error} {
         ::http::log info "Responding."
-        ::http::route $channel $request $routes
+        ::http::route $schannel $request $routes
     } else {
-        puts -nonewline $channel [::http::error-response $error]
+        puts -nonewline $schannel [::http::error-response $error]
     }

+    close $schannel
     close $channel
 }

```

**Comment by dbohdan** [comment]
Implemented in 1417f8283c1916e4bc1b765458c9894df140941e. Try to set the lambda to something like

``` Tcl
{{} {
    tls load_cert ../rootCA.pem
    upvar 1 channel channel
    set channel [$channel tls -server]
}
```

and tell me how it works. Note that for now you will have to modify the proc `::http::serve` itself if you want to have the `break` statement there.

**Comment by dimkr** [comment]
You can close this issue - works great with the LibreSSL bindings from https://github.com/dimkr/jimtcl/tree/libressl.

``` tcl
{{} {
    upvar 1 channel channel
    set channel [$channel ssl -server certificate.pem key.pem]
}
```
